File: app/src/test/java/com/hana053/micropost/ui/pages/main/MainServiceTest.java
package com.hana053.micropost.ui.pages.main;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.hana053.micropost.domain.Micropost;
import com.hana053.micropost.interactors.FeedInteractor;
import com.hana053.micropost.ui.components.micropostlist.PostListAdapter;
import com.hana053.micropost.testing.RobolectricBaseTest;
import com.hana053.micropost.testing.RobolectricDaggerMockRule;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;

import java.util.List;

import rx.Observable;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.when;

public class MainServiceTest extends RobolectricBaseTest {

    @Rule
    public final RobolectricDaggerMockRule rule = new RobolectricDaggerMockRule();

    @Mock
    private FeedInteractor feedInteractor;

    private final PostListAdapter postListAdapter = new PostListAdapter();

    private MainService mainService;

    @Before
    public void setup() {
        when(feedInteractor.loadNextFeed(any())).thenReturn(Observable.empty());
        when(feedInteractor.loadPrevFeed(any())).thenReturn(Observable.empty());
        mainService = new MainServiceImpl(feedInteractor, postListAdapter);
    }

    @Test
    public void shouldLoadNextFeed() {
        final List<Micropost> posts = Stream.of(100, 101)
                .map(id -> new Micropost(id, "", 0, null))
                .collect(Collectors.toList());
        postListAdapter.addAll(0, posts);
        final Observable<List<Micropost>> response = Observable.just(99)
                .map(id -> new Micropost(id, "", 0, null))
                .toList();
        when(feedInteractor.loadNextFeed(100L)).thenReturn(response);

        mainService.loadNextFeed().subscribe();
        advance();

        assertThat(postListAdapter.getItemCount(), is(3));
    }

    @Test
    public void shouldLoadPrevFeed() {
        final List<Micropost> posts = Stream.of(100, 101)
                .map(id -> new Micropost(id, "", 0, null))
                .collect(Collectors.toList());
        postListAdapter.addAll(0, posts);
        final Observable<List<Micropost>> response = Observable.just(102)
                .map(id -> new Micropost(id, "", 0, null))
                .toList();
        when(feedInteractor.loadPrevFeed(101L)).thenReturn(response);

        mainService.loadPrevFeed().subscribe();
        advance();

        assertThat(postListAdapter.getItemCount(), is(3));
    }

}

File: app/src/test/java/com/hana053/micropost/pages/main/MainServiceTest.kt
package com.hana053.micropost.pages.main

import com.hana053.micropost.domain.Micropost
import com.hana053.micropost.service.HttpErrorHandler
import com.hana053.micropost.shared.posts.PostListAdapter
import com.hana053.micropost.testing.RobolectricBaseTest
import com.hana053.micropost.testing.TestMicropost
import com.hana053.myapp.interactors.FeedInteractor
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.verify
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test
import org.mockito.Mockito.`when`
import rx.Observable


class MainServiceTest : RobolectricBaseTest() {

    private val feedInteractor = mock<FeedInteractor>()
    private val httpErrorHandler = mock<HttpErrorHandler>()
    private val postListAdapter = PostListAdapter()
    private val mainService = MainService(
        feedInteractor = feedInteractor,
        httpErrorHandler = httpErrorHandler,
        postListAdapter = postListAdapter
    )

    @Test
    fun shouldLoadNextFeed() {
        `when`(feedInteractor.loadNextFeed(null))
            .doReturn(Observable.just(listOf(TestMicropost)))

        mainService.loadNextFeed().subscribe()
        advance()

        assertThat(postListAdapter.itemCount, `is`(1))
        assertThat(postListAdapter.getFirstItemId(), `is`(TestMicropost.id))
    }

    @Test
    fun shouldLoadMoreNextFeed() {
        postListAdapter.addAll(0, listOf(TestMicropost.copy(id = 1)))
        `when`(feedInteractor.loadNextFeed(1))
            .doReturn(Observable.just(listOf(
                TestMicropost.copy(id = 3),
                TestMicropost.copy(id = 2)
            )))

        mainService.loadNextFeed().subscribe()
        advance()

        assertThat(postListAdapter.itemCount, `is`(3))
        assertThat(postListAdapter.getFirstItemId(), `is`(3L))
        assertThat(postListAdapter.getLastItemId(), `is`(1L))
    }

    @Test
    fun shouldHandleErrorWhenLoadNextFeed() {
        val error = Observable.error<List<Micropost>>(RuntimeException())
        `when`(feedInteractor.loadNextFeed(null)).doReturn(error)

        mainService.loadNextFeed().subscribe()
        advance()

        verify(httpErrorHandler).handleError(any<RuntimeException>())
    }

    @Test
    fun shouldLoadPrevFeed() {
        `when`(feedInteractor.loadPrevFeed(null))
            .doReturn(Observable.just(listOf(TestMicropost)))

        mainService.loadPrevFeed().subscribe()
        advance()

        assertThat(postListAdapter.itemCount, `is`(1))
        assertThat(postListAdapter.getFirstItemId(), `is`(TestMicropost.id))
    }

    @Test
    fun shouldLoadMorePrevFeed() {
        postListAdapter.addAll(0, listOf(TestMicropost.copy(id = 3)))
        `when`(feedInteractor.loadPrevFeed(3))
            .doReturn(Observable.just(listOf(
                TestMicropost.copy(id = 2),
                TestMicropost.copy(id = 1)
            )))

        mainService.loadPrevFeed().subscribe()
        advance()

        assertThat(postListAdapter.itemCount, `is`(3))
        assertThat(postListAdapter.getFirstItemId(), `is`(3L))
        assertThat(postListAdapter.getLastItemId(), `is`(1L))
    }

    @Test
    fun shouldHandleErrorWhenLoadPrevFeed() {
        val error = Observable.error<List<Micropost>>(RuntimeException())
        `when`(feedInteractor.loadPrevFeed(null)).doReturn(error)

        mainService.loadPrevFeed().subscribe()
        advance()

        verify(httpErrorHandler).handleError(any<RuntimeException>())
    }
}